src/main/java/org/folio/circulation/domain/UpdateItem.java

```diff
@@ -25,20 +31,27 @@ public UpdateItem(Clients clients) {
     itemsStorageClient = clients.itemsStorage();
   }
 
-  public CompletableFuture<Result<Item>> onCheckIn(
-    Item item,
-    RequestQueue requestQueue,
-    UUID checkInServicePointId) {
-
+  public CompletableFuture<Result<Item>> onCheckIn(Item item, RequestQueue requestQueue,
+      UUID checkInServicePointId, UUID loggedInUserId, DateTime dateTime) {
     return changeItemOnCheckIn(item, requestQueue, checkInServicePointId)
-      .after(updatedItem -> {
-        if(updatedItem.hasChanged()) {
-          return storeItem(updatedItem);
-        }
-        else {
-          return completedFuture(succeeded(item));
-        }
-      });
+      .next(addLastCheckInProperties(checkInServicePointId, loggedInUserId, dateTime))
+      .after(this::storeItem);
+  }
+
+  private Function<Item, Result<Item>> addLastCheckInProperties(
```

Comment:
Is the rest of the translation to JSON for the item also done here, or is it in the repository?

This method reaches inside the domain object and manipulates the JSON directly. The presence of that JSON is already a compromise, I think we should try to limit access to it to only internally or during specific storage operations.

Reply:
The code has been refactored.

Reply:
I think I’ve not explained myself well.

In the domain, rather than using the JSON representation, use a strongly typed class, and then map it to the JSON during the storage of the record.

Including the raw JSON in the domain was a convenience (because the storage and business logic representations appear similar) to reduce mapping code. That has had significant drawbacks and I think it is a pattern to move away from. The intent is to start to try to separate the domain and the storage module interaction.

Does that make sense?

Reply:
I do understand drawbacks of using JSON representation directly. But that is the existing approach in almost all of the domain classes and it does also bring some benefits, e.g. convinience, etc.   Can you give more inforamation about the significant drawbacks of that apprach specifically in Folio? 
PS: reworking it now

Reply:
It is used in almost all of the domain classes, that is likely because I introduced the style (to avoid the mundane mapping that plagued mod-inventory).

Having two patterns does increase the cognitive load. I don’t think the refactoring away from the incumbent pattern can be done in a single step. Hence I am trying to not increase the amount of rework needed.

I think the primary drawback is that it confuses the domain. When mod-circulation was mostly an intermediary that was ok, now that it has significant logic it adds complexity.

When I’m reading the domain code, I find myself having to think about both the logic and the data representation. It creates coupling between the business logic API representations, the domain itself and the storage representations.

It also undermines the intent for immutability in the domain, as the JSON can be mutated at any point. I think this introduces scope for bugs.

If you think this isn’t a worthwhile tradeoff, then I’m happy for this to become a larger discussion.

Reply:
That was clrear and comprehensive. Thank you.